File: dsp_modules/modules/wah/matlab/wah.m
classdef wah < handle
    % MEMORY
    properties
        input_blocksize; % self explanatory
        sampling_rate;
        output_buffer;
        
        freq_low;
        freq_high;
        freq_delta;     % wah frequency, how many Hz are cycled through pr sample
        freq_center;    % pass band center freq
        Q;              % pass band width, lower value means smaller band
        
        yb_old;        
        yl_old;
                
    end
    
    methods
        %% init
        function obj = wah(input_blocksize, sampling_rate,freq_low,freq_high,freq_delta,Q)
            obj.yb_old = 0;
            obj.yl_old = 0;
            obj.input_blocksize = input_blocksize;
            obj.sampling_rate = sampling_rate;
            obj.freq_low = freq_low;
            obj.freq_high = freq_high;
            obj.freq_delta = freq_delta;
            obj.freq_center = obj.freq_low;
            obj.Q = Q;
            
            obj.output_buffer = zeros(input_blocksize,1);
            
        end
        
        %% apply
        function wah_apply(obj,input_block)
            
            for i = 1:obj.input_blocksize
                
                F1 = 2*sin((pi*obj.freq_center)/obj.sampling_rate);
                
                yh = input_block(i) - obj.yl_old - obj.Q*obj.yb_old;
                yb = F1*yh + obj.yb_old;
                yl = F1*yb + obj.yl_old;
                obj.output_buffer(i) = yb;
                obj.yb_old = yb;
                obj.yl_old = yl;
                
                obj.freq_center = obj.freq_center + obj.freq_delta;
                if obj.freq_center > obj.freq_high % top reached! count down now!
                    obj.freq_center = obj.freq_high;
                    obj.freq_delta = obj.freq_delta * -1;
                end
                
                if obj.freq_center < obj.freq_low % bottom reached! count up now!
                    obj.freq_center = obj.freq_low;
                    obj.freq_delta = obj.freq_delta * -1;
                end
                
                
            end
        end
    end
end

File: dsp_modules/modules/wah/c/wah.c
#include <modules.h>
#include <arm_math.h>

int wah_init(t_wah_data *data, int blocksize, int sampling_rate, float freq_low, float freq_high, float freq_delta, float Q) {
    // error handling
    data->init_ok = 0;
    if (blocksize>WAH_MAX_BLOCKSIZE) return 1;

    // no errors, go on..
    data->blocksize = blocksize;
    data->sampling_rate = sampling_rate;

    data->freq_low = freq_low;
    data->freq_high = freq_high;
    data->freq_delta = freq_delta;
    data->freq_center = data->freq_low;
    data->Q = Q;
 
    wah_reset(data);
    data->init_ok = 1;

    return 0;
}

void wah_reset(t_wah_data *data) {
    arm_fill_f32(0.0f, data->output, data->blocksize);    
    data->yb_old = 0;
    data->yl_old = 0;    
}

void wah_apply(t_wah_data *data, float *input) {    
   
    float F1, yh, yb, yl;

    for(int i = 0; i < data->blocksize; i++)
    {
       
        F1 = 2 * sin((PI*data->freq_center) / data->sampling_rate);

        yh = input[i] - data->yl_old - data->Q*data->yb_old;

        yb = F1*yh + data->yb_old;
        yl = F1*yb + data->yl_old;
        
        data->output[i] = yb;

        data->yb_old = yb;
        data->yl_old = yl;

        data->freq_center = data->freq_center + data->freq_delta;
        if(data->freq_center > data->freq_high) // top reached! count down now!
        {
            data->freq_center = data->freq_high;
            data->freq_delta = data->freq_delta * -1;   
        }

        if (data->freq_center < data->freq_low) // bottom reached! count up now!
        {
            data->freq_center = data->freq_low;
            data->freq_delta = data->freq_delta * -1;
        }    
            
    }

}